openedx/core/djangoapps/content/course_overviews/models.py

```diff
@@ -0,0 +1,251 @@
+        try:
+            course_overview = CourseOverview.objects.get(id=course_id)
+            # Cache hit! Just return the overview
+        except CourseOverview.DoesNotExist:
+            # Cache miss. Load entire course and create a CourseOverview from it
+            course = modulestore().get_course(course_id)
+            if course:
+                course_overview = CourseOverview._create_from_course(course)
+                course_overview.save()  # Save new overview to the cache
+        return course_overview
+
+    def __repr__(self):
+        """
+        Returns a simple string representation of this object for debugging.
+
+        Example return value: CourseOverview(location=course-v1://edX+DemoX.1+2014)
+        """
+        return "{}(location={})".format(self.__class__.__name__, self.location)
+
+    def __str__(self):
+        """
+        Returns a string representation of this object suitable for a user to see.
+        """
+        return unicode(self.id)
+
+    def clean_id(self, padding_char='='):
+        """
+        Returns a unique deterministic base32-encoded ID for the course.
+
+        Arguments:
+            padding_char (str): Character used for padding at end of base-32
+                                -encoded string, defaulting to '='
+        """
+        return course_metadata_utils.clean_id(self.location, padding_char)
+
+    @property
+    def location(self):
+        """
+        Returns the UsageKey of this course.
+
+        UsageKeyField has a strange behavior where it fails to parse the "run"
+        of a course out of the serialized form of a Mongo Draft UsageKey. This
+        method is a wrapper around _location attribute that fixes the problem
+        by calling map_into_course, which restores the run attribute.
+        """
+        return self._location.map_into_course(self.id)  # pylint: disable=no-member
+
+    @property
+    def number(self):
+        """
+        Returns this course's number.
+        """
+        return course_metadata_utils.course_number(self.location)
+
+    @property
+    def url_name(self):
+        """
+        Returns this course's URL name.
+        """
+        return course_metadata_utils.course_url_name(self.location)
+
+    @property
+    def display_name_with_default(self):
+        """
+        Return reasonable display name for the course.
+        """
+        return course_metadata_utils.display_name_with_default(self)
+
+    def has_started(self):
+        """
+        Returns whether the the course has started.
+        """
+        return course_metadata_utils.has_course_started(self.start)
+
+    def has_ended(self):
+        """
+        Returns whether the course has ended.
+        """
+        return course_metadata_utils.has_course_ended(self.end)
+
+    def start_datetime_text(self, format_string="SHORT_DATE"):
+        """
+        Returns the desired text corresponding the course's start date and time in UTC.  Prefers .advertised_start,
```

Comment:
FYI, although our code char limit is 120, docstrings are suggested to be < 80 char.

https://github.com/edx/edx-platform/wiki/Python-Guidelines